### ISSUE TITLE
backend install issue

### ISSUE THREAD

**Issue by zekdemir** [GChristensen/scrapyard#89]
Hello! Scrapyard is great. Lately I've tried to install the backend on several different linux systems but it always fails saying an error like externally-managed-environment and complain about not finding an environment as below

ERROR: Could not find a version that satisfies the requirement installer.py (from versions: none)                                                                                                                 
ERROR: No matching distribution found for installer.py 

I tried everything with python versions above 3.7 as required yet it is never successful. Neither "install.sh" nor "installer.py" works. Might it be that the backend install script needs an update? It didn't happen before.

Thanks!

**Comment by GChristensen** [comment]
Hi, 

More context is needed to find out what is happening. The errors suggest that pip may be missing as it may happen on Ubuntu. The installation script relies on standard Python tools, including pip.

**Comment by zekdemir** [comment]
I must admit I am not very knowledgeable either on linux or python. Yet I tried all suggestions by python on command line and also searched internet a lot yet nothing I found worked. Yes, I tried pip and also pipx. It always comes down to a virtual environment issue which I am, with my current level on linux and python, unable to resolve. I guess it works with some older versions of python as I was, only, able to install it on Slackware 15. When I wanted to change my linux system, it didn't work on 2 debian systems and 1 artix.

Anyhow, I am happy for your prompt reply. I guess I am just gonna have to find a replacement till it starts working in some way again :) Zotero looks like an alternative to scrapbook which is also integrated into firefox easily but transfering a large collection into it may prove hard.

Anyway, I'll be watching for future updates. Thanks for providing Scrapyard!

**Comment by GChristensen** [comment]
You may post your version of Linux, the version of python (could be obtained with `python3 --version` command) and detailed console error text, and if `setuptools` presents in the output of the `pip list` command. It would be easier to fix that.

**Comment by zekdemir** [comment]
It is Puppy Linux. Debian Dog more precisely.

This is the result with "./install.sh":

puppy@live:~/Downloads/scrapyard-backend-2.1.2$ python3 --version
Python 3.11.2
puppy@live:~/Downloads/scrapyard-backend-2.1.2$ ./install.sh
error: externally-managed-environment

× This environment is externally managed
╰─> To install Python packages system-wide, try apt install
    python3-xyz, where xyz is the package you are trying to
    install.
    
    If you wish to install a non-Debian-packaged Python package,
    create a virtual environment using python3 -m venv path/to/venv.
    Then use path/to/venv/bin/python and path/to/venv/bin/pip. Make
    sure you have python3-full installed.
    
    If you wish to install a non-Debian packaged Python application,
    it may be easiest to use pipx install xyz, which will manage a
    virtual environment for you. Make sure you have pipx installed.
    
    See /usr/share/doc/python3.11/README.venv for more information.

note: If you believe this is a mistake, please contact your Python installation or OS distribution provider. You can override this, at the risk of breaking your Python installation or OS, by passing --break-system-packages.
hint: See PEP 668 for the detailed specification.
Traceback (most recent call last):
  File "/home/puppy/Downloads/scrapyard-backend-2.1.2/./scrapyard_backend/installer.py", line 61, in <module>
    subprocess.check_call([sys.executable, "-m", "pip", "install", package_path, "--user"])
  File "/usr/lib/python3.11/subprocess.py", line 413, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['/usr/bin/python3', '-m', 'pip', 'install', '/home/puppy/Downloads/scrapyard-backend-2.1.2/scrapyard_backend', '--user']' returned non-zero exit status 1.
puppy@live:~/Downloads/scrapyard-backend-2.1.2$

**Comment by zekdemir** [comment]
And the result with "python3 -m pip install installer.py" as suggested above:

puppy@live:~/Downloads/scrapyard-backend-2.1.2$ cd scrapyard_backend/
puppy@live:~/Downloads/scrapyard-backend-2.1.2/scrapyard_backend$ python3 -m pip install installer.py
error: externally-managed-environment

× This environment is externally managed
╰─> To install Python packages system-wide, try apt install
    python3-xyz, where xyz is the package you are trying to
    install.
    
    If you wish to install a non-Debian-packaged Python package,
    create a virtual environment using python3 -m venv path/to/venv.
    Then use path/to/venv/bin/python and path/to/venv/bin/pip. Make
    sure you have python3-full installed.
    
    If you wish to install a non-Debian packaged Python application,
    it may be easiest to use pipx install xyz, which will manage a
    virtual environment for you. Make sure you have pipx installed.
    
    See /usr/share/doc/python3.11/README.venv for more information.

note: If you believe this is a mistake, please contact your Python installation or OS distribution provider. You can override this, at the risk of breaking your Python installation or OS, by passing --break-system-packages.
hint: See PEP 668 for the detailed specification.
puppy@live:~/Downloads/scrapyard-backend-2.1.2/scrapyard_backend$

**Comment by GChristensen** [comment]
Thanks, I've grasped the nature of the error. Changes in the global Python environment are forbidden in this distro. I'll think how to install the backed at venv. But this may be not easy/fast fix.

**Comment by zekdemir** [comment]
It didn't happen in Slackware (Porteus, which is a portable linux, similar to Puppy) interestingly. I've also encountered very similar issues on my search on the internet with different people with different python apps/scripts. I'll be waiting for a possible solution hopefully. Thanks for your efforts!

**Comment by GChristensen** [comment]
At the first sight, the fix is easy. In this archive: [installer.zip](https://github.com/user-attachments/files/17138941/installer.zip) is the fixed `installer.py` file to replace the `scrapyard_backend/installer.py` from the archive. The backend is installed in venv. It works on my Linux.
The fix need some testing, but the solution is viable.

**Comment by zekdemir** [comment]
Thanks! It worked ok with "./install.sh" command now with some warning:

DEPRECATION: scrapyard-backend is being installed using the legacy 'setup.py install' method, because it does not have a 'pyproject.toml' and the 'wheel' package is not installed. pip 23.1 will enforce this behaviour change. A possible replacement is to enable the '--use-pep517' option. Discussion can be found at https://github.com/pypa/pip/issues/8559

While waiting, I tested it again with "./install.sh" command on Porteus and FatDog (which is also Porteus-Slackware like source-based system) and it installed with zero issues, downloaded everything required. I also tried the same command with Arch Linux and Ubuntu but it didn't worked and produced the same errors as in posts above. Porteus was with python 3.9 and FatDog with python 3.11. Arch and Ubuntu were with python 3.12 if I can remember correctly. By the way, as  you probably already know, the script worked on ones without systemd init and didn't work (or I couldn't get it to work) on ones with systemd init, if it is related to that at all.

I will try it with the new script on those two again. Thank you very much for your precious time!
